src/metabase/api/cloud_migration.clj

```diff
@@ -0,0 +1,161 @@
+
+    ;; Users need to login, make queries, and we need need to audit them.
+    :model/Session
+    :model/User
+    :model/LoginHistory
+    :model/AuditLog
+    :model/QueryExecution
+    :model/ViewLog
+
+    ;; TODO: Came up in tests, check with core folks
+    :setting
+    :permissions_group_membership})
+
+;; Block write calls to most tables in read-only mode.
+(methodical/defmethod t2.pipeline/build :before [#_query-type     :toucan.statement-type/DML
+                                                 #_model          :default
+                                                 #_resolved-query :default]
+  [_query-type model _parsed-args resolved-query]
+  (when (and (read-only-mode)
+             (not (read-only-mode-exceptions model)))
+    (throw (ex-info "Metabase is in read-only-mode mode!" {})))
+  resolved-query)
+
+
+;; Helpers
+
+(def ^:private terminal-states
+  #{:done :error :cancelled})
+
+(defn- cluster?
+  []
+  (>= (t2/count 'QRTZ_SCHEDULER_STATE) 2))
+
+(defn- set-progress
+  [id state progress]
+  (when (= 0 (t2/update! CloudMigration :id id :state [:not-in terminal-states]
+                         {:state state :progress progress}))
+    (throw (ex-info "Cannot update migration in terminal state" {:terminal true}))))
+
+(defn- migrate! [{:keys [id]}]
+  (try
+    ;; Set read-only-mode
+    (set-progress id :setup 1)
+    (read-only-mode! true)
+    (when (cluster?)
+      ;; Wait for read-only-mode to propagate to other cluster instances.
+      (Thread/sleep (int (* 1.5 setting.cache/cache-update-check-interval-ms))))
+
+    ;; Start dumping the h2 backup
+    (set-progress id :dump 20)
+    (dump-to-h2/dump-to-h2! h2-file)
+    (read-only-mode! false)
+
+    ;; Upload dump to store
+    (set-progress id :upload 50)
+    ;; TODO: upload
+    ;; TODO: track upload progress
+
+    ;; All done
+    (set-progress id :done 100)
+    (catch Exception e
+      (when-not (-> e ex-data :terminal)
+        (t2/update! :id id :state :error)
+        (log/error e "Error performing migration")))
+    (finally
+      (read-only-mode! false)
+      (io/delete-file h2-file :silently))))
+
+(defn- get-store-migration
+  "Calls Store and returns {:external_id ,,, :upload-url ,,,}."
+  []
+  (-> metabase-store-migration-url
+      (http/post {:form-params  {:local_mb_version (config/mb-version-info :tag)}
+                  :content-type :json})
+      :body
+      (json/parse-string keyword)
+      (select-keys [:id :upload_url])
+      (set/rename-keys {:id :external_id})))
+
+
+;; Endpoints
+
+(api/defendpoint POST "/"
+  "Initiate a new cloud migration."
+  [_]
+  (api/check-superuser)
+  (if (t2/select-one CloudMigration :state [:not-in terminal-states])
+    {:status 409 :body "There's an ongoing migration already."}
+    (try
+      (let [cloud-migration (t2/insert-returning-instance! CloudMigration (get-store-migration))]
+        (future (migrate! cloud-migration))
+        cloud-migration)
+      (catch Exception e
```

Comment:
this `catch` is probably doing less work than you think since the `migrate!` call is in a `future`

Reply:
It's for the error status codes in `get-store-migration`. TBH I wasn't very happy with putting it here. I should put it in `get-store-migration`, but then I still have somehow say "well if that one returns an error, return that instead".